kahypar/datastructure/graph.h

```diff
@@ -25,826 +25,729 @@
+    _weighted_degree(_num_nodes, 0.0L),
+    _cluster_id(_num_nodes),
+    _cluster_size(_num_nodes, 1),
+    _incident_cluster_weight(_num_nodes, IncidentClusterWeight(0, 0.0L)),
+    _incident_cluster_weight_position(_num_nodes),
+    _hypernode_mapping(hypergraph.initialNumNodes() + hypergraph.initialNumEdges(), kInvalidNode) {
+    std::iota(_shuffle_nodes.begin(), _shuffle_nodes.end(), 0);
+    std::iota(_cluster_id.begin(), _cluster_id.end(), 0);
+    if (config.preprocessing.louvain_community_detection.use_bipartite_graph) {
+      const auto degreeWeight = [&](const Hypergraph& hg,
+                                    const HyperedgeID he,
+                                    const HypernodeID hn) {
+                                  return (static_cast<EdgeWeight>(hg.edgeWeight(he)) *
+                                          static_cast<EdgeWeight>(hg.nodeDegree(hn))) /
+                                         static_cast<EdgeWeight>(hg.edgeSize(he));
+                                };
+      const auto uniformWeight = [&](const Hypergraph& hg,
+                                     const HyperedgeID he,
+                                     const HypernodeID) {
+                                   return static_cast<EdgeWeight>(hg.edgeWeight(he));
+                                 };
+
+      switch (config.preprocessing.louvain_community_detection.edge_weight) {
+        case LouvainEdgeWeight::degree:
+          constructBipartiteGraph(hypergraph, degreeWeight);
+          break;
+        case LouvainEdgeWeight::non_uniform:
+          constructBipartiteGraph(hypergraph,
+                                  [&](const Hypergraph& hg,
+                                      const HyperedgeID he,
+                                      const HypernodeID) {
+              return static_cast<EdgeWeight>(hg.edgeWeight(he)) /
+              static_cast<EdgeWeight>(hg.edgeSize(he));
+            });
+
+          break;
+        case LouvainEdgeWeight::uniform:
+          constructBipartiteGraph(hypergraph, uniformWeight);
+          break;
+        case LouvainEdgeWeight::hybrid:
+          const double density = static_cast<double>(hypergraph.initialNumEdges()) /
+                                 static_cast<double>(hypergraph.initialNumNodes());
+          if (density < 0.75) {
+            constructBipartiteGraph(hypergraph, degreeWeight);
+          } else if (density >= 0.75 && density <= 1.25) {
+            constructBipartiteGraph(hypergraph, uniformWeight);
+          } else {
+            constructBipartiteGraph(hypergraph, uniformWeight);
+          }
+          break;
+      }
     }
-    
-    ClusterID hypernodeClusterID(const HypernodeID hn) const {
-        ASSERT(_hypernodeMapping[hn] != INVALID_NODE, "Hypernode " << hn << " isn't part of the current graph!");
-        return _cluster_id[_hypernodeMapping[hn]];
+  }
+
+  Graph(const std::vector<NodeID>& adj_array, const std::vector<Edge>& edges) :
```

Comment:
No, it's also needed in the karate club test case, since you build the adjacency structure there.